Fix: return empty table if no stub-exits for getExitStubs functions

SlySven opened the pull request:
The existing code produces a nil+warning message if there are no exit stubs for a room when `getExitStubs(...)` OR `getExitStubs1(...)` are called on it. This is unfortunate because the only way to find out whether a room has any exit stubs in a script is by using those functions.

Unlike the draft #6069 this does not produce a `nil` *without* a message in just this "problem" case, instead it returns an empty table, so the table can be stored locally and examined in both the case where there are and are not exit stubs. I feel this is a cleaner solution because there are other run-time problem situations that also will produce a `nil` and a message - which that draft doesn't really resolve IMHO.

~~To enable this fix (and thus to change the behaviour away from how a previous script/package might be working to examine the precise error message and thus not be entirely backward compatible) it is necessary to set the global boolean variable `noWarningForNoStubExits` to `true` - this has the nice side-effect that a script/package could test for the existence of this variable to test whether the fix is available!~~

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

*Edited 2022/06/18 to reflect revisions of [cb4e0e4](https://github.com/Mudlet/Mudlet/pull/6081/commits/cb4e0e4e01dadd947708058b41ff69e9f111865e)*

SlySven committed:
Fix: return empty table if no stub-exits for getExitStubs functions

The existing code produces a nil+warning message if there are no exit stubs
for a room when `getExitStubs(...)` OR `getExitStubs1(...)` are called on
it. This is unfortunate because the only way to find out whether a room has
any exit stubs in a script is by using those functions.

Unlike the draft #6069 this does not produce a `nil` *without* a message
in just this "problem" case, instead it just returns an empty table, so the
table can be stored locally and examined in both the case where there are
and are not exit stubs. I feel this is a cleaner solution because there
are other run-time problem situations that also will produce a `nil` and
a message - which that draft doesn't really resolve IMHO.

To enable this fix (and thus to change the behaviour away from how a
previous script/package might be working to examine the precise error
message and thus not be entirely backward compatible) it is necessary to
set the global boolean variable `noWarningForNoStubExits` to `true` - this
has the nice side-effect that a script/package could test for the existance
of this variable to test whether the fix is available!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

keneanung reviewed (commented):
I don't like the global config-like variable. I get where you are coming from and I think it has merit to think about this breaking change. But for me, this is more of a bug than a changed feature.  I'm open to discussion and wouldn't completely block this solution, but we should try to find a better one first.

vadi2 commented:
I'm with @keneanung on this one, change the return for no stubs to an empty table instead of nil (if we're not doing a breaking change and causing grief by doing so) and stay out of global flags like this.

SlySven committed:
Merge branch 'development' into Fix_returnEmptyTableIfNoStubsFor_getExitStubsFunctions

SlySven committed:
Revise: getExit{|1}(...) return empty table unconditionally on no stubs

This changes from previous behaviour of a `nil` plus an error message when
there were no stubs to report and of just returning a `nil` in even older
versions of Mudlet. Technically this could be held to be breaking backwards
behaviour but was suggested as desirable by both @keneanung and @vadi2.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>
